Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp

```diff
@@ -99,16 +101,312 @@ ErrorOr<VP8Header> decode_webp_chunk_VP8_header(ReadonlyBytes vp8_data)
+ErrorOr<void> decode_VP8_frame_header_coefficient_probabilities(BooleanDecoder&, CoefficientProbabilities);
+
+// https://datatracker.ietf.org/doc/html/rfc6386#section-15 "Loop Filter"
+// "The first is a flag (filter_type) selecting the type of filter (normal or simple)"
+enum class FilterType {
+    Normal = 0,
+    Simple = 1,
+};
+
+// https://datatracker.ietf.org/doc/html/rfc6386#section-19.2 "Frame Header"
+struct FrameHeader {
+    ColorSpaceAndPixelType color_space {};
+    ClampingSpecification clamping_type {};
+
+    bool is_segmentation_enabled {};
+    Segmentation segmentation {};
+
+    FilterType filter_type {};
+    u8 loop_filter_level {};
+    u8 sharpness_level {};
+    LoopFilterAdjustment loop_filter_adjustment {};
+
+    u8 number_of_dct_partitions {};
+
+    QuantizationIndices quantization_indices {};
+
+    CoefficientProbabilities coefficient_probabilities;
+
+    bool enable_skipping_of_metablocks_containing_only_zero_coefficients {};
+    u8 probability_skip_false;
+};
+
+ErrorOr<FrameHeader> decode_VP8_frame_header(BooleanDecoder& decoder)
+{
+    // https://datatracker.ietf.org/doc/html/rfc6386#section-19.2 "Frame Header"
+    FrameHeader header;
+
+    // In the VP8 spec, this is in an `if (key_frames)`, but webp files only have key frames.
+    header.color_space = ColorSpaceAndPixelType { TRY(L(1)) };
+    header.clamping_type = ClampingSpecification { TRY(L(1)) };
+    dbgln_if(WEBP_DEBUG, "color_space {} clamping_type {}", (int)header.color_space, (int)header.clamping_type);
+
+    // https://datatracker.ietf.org/doc/html/rfc6386#section-9.3 "Segment-Based Adjustments"
+    header.is_segmentation_enabled = TRY(L(1));
+    dbgln_if(WEBP_DEBUG, "segmentation_enabled {}", header.is_segmentation_enabled);
+
+    if (header.is_segmentation_enabled)
+        header.segmentation = TRY(decode_VP8_frame_header_segmentation(decoder));
+
+    header.filter_type = FilterType { TRY(L(1)) };
+    header.loop_filter_level = TRY(L(6));
+    header.sharpness_level = TRY(L(3));
+    dbgln_if(WEBP_DEBUG, "filter_type {} loop_filter_level {} sharpness_level {}", (int)header.filter_type, header.loop_filter_level, header.sharpness_level);
+
+    header.loop_filter_adjustment = TRY(decode_VP8_frame_header_loop_filter_adjustment(decoder));
+
+    u8 log2_nbr_of_dct_partitions = TRY(L(2));
+    dbgln_if(WEBP_DEBUG, "log2_nbr_of_dct_partitions {}", log2_nbr_of_dct_partitions);
+    header.number_of_dct_partitions = 1 << log2_nbr_of_dct_partitions;
+
+    header.quantization_indices = TRY(decode_VP8_frame_header_quantization_indices(decoder));
+
+    // In the VP8 spec, this is in an `if (key_frames)` followed by a lengthy `else`, but webp files only have key frames.
+    u8 refresh_entropy_probs = TRY(L(1)); // Has no effect in webp files.
+    dbgln_if(WEBP_DEBUG, "refresh_entropy_probs {}", refresh_entropy_probs);
+
+    memcpy(header.coefficient_probabilities, default_coeff_probs, sizeof(header.coefficient_probabilities));
```

Comment:
Any reason the compiler-generated copy ctor for this POD struct isn't sufficient? I think elsewhere we've avoided memcpy for that usage. Or that we couldn't `= default_coeff_probs` in the designated initializer for the FrameHeader struct.

Reply:
it don't build:

```
WebPLoaderLossy.cpp:239:38: error: array type 'Gfx::(anonymous namespace)::CoefficientProbabilities' (aka 'unsigned char[4][8][3][11]') is not assignable
    header.coefficient_probabilities = default_coeff_probs;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```

(repro:

```sh
% cat test.cc
using t = int[2][2];

struct s {
    t k;
};

int main() {
    int a[2][2] = { { 1, 2}, {3, 4} };

    s k;
    k.k = a;
}
% clang -c test.cc -std=c++11
test.cc:11:9: error: array type 't' (aka 'int[2][2]') is not assignable
    k.k = a;
    ~~~ ^
```

Reply:
(ie it ain't a struct but an array)

Reply:
ah, oki. Well, we could "prefer AK::Array to C-style arrays" but I don't think that one is actually written down (and I don't care *that* much myself).